### DISCUSSION TITLE
Aqualink Z4

### DISCUSSION THREAD

**Discussion by zeropoint46** [aqualinkd/AqualinkD#347] [Q&A]
I have a jandy z4 with iaqualink, is this setup supported at all? thanks.

**Comment by sfeakes** [comment]
Short answer, No.
I’ve never had a Z4 to play with, and no one (to my knoladge) has tried AqualinkD against a Z4.  The Z4 looks like it’s a PDA panel with a nicer interface on the panel, ie very limited functionality.  There is a chance that AqualinkD might be able to communicate with the Z4. But according to the Z4 manual it only supports one UI device, the iAqualink WiFi adapter. If that is the case, then you would need to disconnect the iAqualink device for AqualinkD to work.

**Reply by zeropoint46** [reply]
Got it, I'm not ready to tackle that project quite yet, but I might try and remove the iaqualink and connect the rs485 adapter and fire up aqualinkd. At least give you some logs if you're interested. I also have an LX heater, and I saw somewhere in the docs or wiki that you might need some logs from that as well. Thanks again for the quick response.

**Reply by sfeakes** [reply]
The serial_logger included with AqualinkD will give you ALL the information you need, and there would be no need to disconnect the iAqualink wifi device to do that.  So if you do get time / opportunity, wire up an RS485 device (with iAqualink device attached) and run serial_logger in debug mode.  if you post the output, I'll be able to tell you right away how much / what we could do to get it to work.

**Reply by zeropoint46** [reply]
ohh thats fantastic, thanks. should I post here? or open a new discussion?

**Reply by sfeakes** [reply]
Post it here would be good, that way we have context.

**Reply by zeropoint46** [reply]
oh, is there any documentation for the serial logger? anything I should do once I start logging? should I issue commands on the control panel?

**Reply by zeropoint46** [reply]
found the docs

**Comment by zeropoint46** [comment]
got the output:

```
AqualinkD serial_logger V2.6
Notice:  RS Serial: Port /dev/ttyUSB1 low latency mode is NOT set, resetting to low latency!
Notice:  Serial Log:Logging serial information!
Please wait.|

Notice:  Serial Log:RS485 interface received 800 packets in 62 seconds ( 12.90 Msg/Sec)
Notice:  Serial Log:Jandy Control Panel Model   : 
Notice:  Serial Log:Jandy Control Panel Version : 
Notice:  Serial Log:Jandy ID's found
Notice:  Serial Log:ID 0x33 is in use    <-- Aqualink (iAqualink / Touch)
Notice:  Serial Log:ID 0xc8 is in use    <-- Unknown Device
Notice:  Serial Log:ID 0xa3 is in use    <-- iAqualink 2.0
Notice:  Serial Log:ID 0x68 is in use    <-- LXi / LRZ Heater
Notice:  Serial Log:ID 0x78 is in use    <-- Jandy VSP ePump
Notice:  Serial Log:ID 0x60 is not used  <-- can use for Aqualinkd (PDA mode only)
Notice:  Serial Log:

Notice:  Serial Log:Suggested aqualinkd.conf values
Notice:  Serial Log:-------------------------
Notice:  Serial Log:device_id = 0x60
Notice:  Serial Log:read_RS485_ePump = yes
Notice:  Serial Log:read_RS485_JXi = yes
Notice:  Serial Log:-------------------------
```

**Reply by zeropoint46** [reply]
Reading the wiki/docs more and better, I realize I made a few mistakes the first time. I have now run logger in debug mode and have extended the capture to 2000 packets instead. I have attached the log here.
[jandy_debug.log](https://github.com/user-attachments/files/16985907/jandy_debug.log)

**Comment by sfeakes** [comment]
That output looks very promising, I think this might be a lot easier than I originally expected.  The Z4 panel looks like they used the PDA Rev 6.0 chip as the base.  So this should work if you put AqualinkD in PDA mode.  PDA mode is a lot slower than RS mode, but should get the job done.  If you edit `aqualinkd.conf` and use the setting suggested, (and modify the buttons per wiki) you should be good to go. 
--EDIT--
You should NOT have to disconnect the wifi adapter.  AqualinkD will communicate with the panel on id 0x60 and not interfere with anything else.

**Reply by zeropoint46** [reply]
that sounds great, what about the possibility of removing the iaqualink, using ID 0x33 and not being in PDA mode?

**Reply by sfeakes** [reply]
The WiFi adapter is using 0x33 and 0xa3. At the moment AqualinkD can only use 0x33 as I have not fully decoded iAqualink 2.0 protocol 0xa3.
To put AqualinkD into RS mode, the panel needs to support a lot more protocols than your Z4 panel does.  But in the future I hope to create something nearly as fast using both 0x33 and 0xa3.  But ultimately the Z4 and PDA panels will always be crippled over an RS panel.
So at the moment disconnecting the WiFi adapter and letting AqualinkD use 0x33 isn’t going to buy you much/anything.  Hopefully in the future it will be a lot better

**Reply by zeropoint46** [reply]
makes sense, I'll give the pda route a shot and report back, maybe it'll help others. Also, you mentioned "modify the buttons per wiki" but I'm not sure I can find that section, can you point me to what you mean by that? thanks.

**Reply by zeropoint46** [reply]
I think I found it near the end of configuring aqualinkd section.

**Reply by sfeakes** [reply]
With PDA mode the button names HAVE to be identical to how you named the buttons on your panel.  So this would what you have named the "Filter Pump" "Spa" "Aux1" "Aux2" "Aux3" for the Z4 I believe.   In the devices page on the iAqualink web UI, the ones highlighted as ON.  
![Screenshot 2024-09-12 at 6 51 44 PM](https://github.com/user-attachments/assets/505696d9-2c70-4a0a-9b25-032cc16146ea)

**Reply by zeropoint46** [reply]
I had a chance to test more this weekend. I think I might be out of luck now that I've learned a bit more, but I'll dump the logs here just in case.

```
Notice:  AqualinkD: Aqualink Daemon v2.4.1 (Dev)
Notice:  AqualinkD: Panel set to PDA-6 Combo Pool/Spa
Notice:  AqualinkD: Config log_level         = 5
Notice:  AqualinkD: Config device_id         = 0x60
Notice:  AqualinkD: Config rssa_device_id    = 0x00
Notice:  AqualinkD: Config extra_device_id   = 0x00
Notice:  AqualinkD: Config extra_device_prog = NO
Notice:  AqualinkD: Config serial_port       = /dev/ttyUSB0
Notice:  AqualinkD: Config socket_port       = 80
Notice:  AqualinkD: Config web_directory     = /var/www/aqualinkd/
Notice:  AqualinkD: Config use_aux_labels    = NO
Notice:  AqualinkD: Config override frz prot = NO
Notice:  AqualinkD: Config mqtt server       = (null)
Notice:  AqualinkD: Config mqtt DZ sub topic = (null)
Notice:  AqualinkD: Config mqtt DZ pub topic = (null)
Notice:  AqualinkD: Config mqtt AQ topic     = (null)
Notice:  AqualinkD: Config mqtt user         = (null)
Notice:  AqualinkD: Config mqtt passwd       = (null)
Notice:  AqualinkD: Config mqtt timed update = NO
Notice:  AqualinkD: Config mqtt HA dis topic = (null)
Notice:  AqualinkD: Config mqtt ID           = aqualinkd_74782705
Notice:  AqualinkD: Config PDA Mode          = YES
Notice:  AqualinkD: Config PDA Sleep Mode    = YES
Notice:  AqualinkD: Config force SWG         = NO
Notice:  AqualinkD: Config force PS setpoint = NO
Notice:  AqualinkD: Config force Freeze Prot = NO
Notice:  AqualinkD: Config deamonize         = NO
Notice:  AqualinkD: Config enable scheduler  = YES
Notice:  AqualinkD: Config light_pgm_mode    = 0.00
Notice:  AqualinkD: Debug RS485 protocol     = NO
Notice:  AqualinkD: Debug RS485 protocol raw = NO
Notice:  AqualinkD: Display warnings in web  = NO
Notice:  AqualinkD: Keep panel time in sync  = YES
Notice:  AqualinkD: Read SWG direct          = NO
Notice:  AqualinkD: Read ePump direct        = YES
Notice:  AqualinkD: Read vsfPump direct      = NO
Notice:  AqualinkD: Read JXi heater direct   = YES
Notice:  AqualinkD: Read LX heater direct    = NO
Notice:  AqualinkD: Read Chem Feeder direct  = NO
Notice:  AqualinkD: Serial FTDI low latency  = YES
Notice:  AqualinkD: RS485 Frame delay        = 4ms
Notice:  AqualinkD: Config BTN Filter_Pump   = label Filter Pump     | VSP ID 0x78 | PumpID 1 | 
Notice:  AqualinkD: Config BTN Spa_Mode      = label Spa             | 
Notice:  AqualinkD: Config BTN Aux_1         = label Pool Heat       | 
Notice:  AqualinkD: Config BTN Aux_2         = label Spa Heat        | 
Notice:  AqualinkD: Config BTN Aux_3         = label Lights          | 
Notice:  AqualinkD: Config BTN Aux_4         = label Cleaner         | 
Notice:  AqualinkD: Config BTN Aux_5         = label Aux3            | 
Notice:  AqualinkD: Config BTN Pool_Heater   = label All Off         | 
Notice:  AqualinkD: Config BTN Spa_Heater    = label SPA HEAT        | 
Notice:  AqualinkD: Config BTN Solar_Heater  = label EXTRA AUX       | 
Notice:  NetService:Starting network services thread
Notice:  NetService:Starting web server on port 80
Notice:  NetService:Starting MQTT client to (null)
Notice:  RS Serial: Port /dev/ttyUSB0 low latency mode is set
Notice:  AqualinkD: Listening to Aqualink RS8 on serial port: /dev/ttyUSB0
Notice:  AqualinkD: Waiting for Control Panel probe
Notice:  AqualinkD: Got message but no probe on '0x60', did we start too soon? (waiting for probe)
Error:   AqualinkD: No probe on '0x60', giving up! (please check config)
```

so I think while maybe some parts of whatever zodiac harvested to make the Z4 and turn it into whatever zombie that it is make aqualinkd think that PDA support is available. But looking at this output, I guess it never gets a probe from the equipment? not sure, just hypothesizing now. Either way, any input would be greatly appreciated. thanks.

**Reply by zeropoint46** [reply]
also I did try a few more time and got some of this:
```
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial: Read  Jandy   packet BAD PACKET To 0xc8 of type         iAq Poll | HEX: 0x10|0x02|0xc8|0x30|0xa8|0x0e|0x34|0x10|0x02|0xa3|0x53|0x08|0x10|0x03| 
Notice:  AqualinkD: Got message but no probe on '0x60', did we start too soon? (waiting for probe)
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial: Read  Jandy   packet BAD PACKET To 0x78 of type   Unknown '0xfb' | HEX: 0x10|0x02|0x78|0xfb|0x10|0x02|0xa3|0x53|0x08|0x10|0x03| 
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial: Read  Jandy   packet BAD PACKET To 0x10 of type           Status | HEX: 0x10|0x02|0x10|0x02|0xff|0xba|0x34|0x10|0x02|0x33|0x30|0x75|0x10|0x03| 
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial: Read  Jandy   packet BAD PACKET To 0x00 of type     iAq pMessage | HEX: 0x10|0x02|0x00|0x25|0x00|0x00|0x01|0x00|0x80|0x14|0x03|0xca|0x03|0x00|0x80|0x02|0x0f|0x14|0x10|0x32|0x06|0x00|0x00|0x00|0x00|0x00|0x00|0xf8|0x34|0x10|0x02|0x78|0x43|0xcd|0x10|0x03| 
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial: Read  Jandy   packet BAD PACKET To 0x78 of type   Unknown '0x43' | HEX: 0x10|0x02|0x78|0x43|0x85|0x10|0x03| 
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial: Read  Jandy   packet BAD PACKET To 0x33 of type         iAq Poll | HEX: 0x10|0x02|0x33|0x30|0x75|0xf0|0x00|0x00|0x10|0x02|0x00|0x01|0x00|0x00|0x13|0x10|0x03| 
```
so maybe I wired it up poorly? however I'm not sure if those are just red herrings.

**Reply by sfeakes** [reply]
I would check the wiring. It looks like you have some issues there.

On the first set of logs `No probe on '0x60', giving up!`  it could be simply a slow start on the Z4.  You got a probe message with the serial_logger, so  I'm guessing you may need to start it a few times once you clean up the wiring issue.

**Reply by zeropoint46** [reply]
Wait, do I need to power off the z4 first and power it on while connected?

**Reply by sfeakes** [reply]
No, should be no need to cycle power on Z4

**Reply by zeropoint46** [reply]
k, cleaned up wiring, have it solidly hardwired in and tightened down on both ends, getting no checksum errors but still get the same:
```
Notice:  AqualinkD: Waiting for Control Panel probe
Notice:  AqualinkD: Got message but no probe on '0x60', did we start too soon? (waiting for probe)
Error:   AqualinkD: No probe on '0x60', giving up! (please check config)
```
not sure what else I can do now tho.

**Reply by zeropoint46** [reply]
not sure whats going on, but I think the little Z4 control panel in the front of the unit is like a PDA or something. it's messing with intercepting the 0x60 probe or something. Not sure, but I feel like this isn't gonna work.

**Reply by sfeakes** [reply]
When you ran serial_logger previously, you got a probe on 0x60, and that's all AqualinkD is looking for to start.  The below message means serial_logger saw a probe message on ID 0x60. (A probe message means the Panel is looking for a device at that ID, and nothing replied to that message, meaning there is no device at that ID). 
```Notice:  Serial Log:ID 0x60 is not used  <-- can use for Aqualinkd (PDA mode only)```
Run serial_logger again and see if you get the above message?

If serial_logger sees something replying to the ID, then it will mark that ID as in use, as you can see from your other outputs.
```
Notice:  Serial Log:ID 0x33 is in use    <-- Aqualink (iAqualink / Touch)
Notice:  Serial Log:ID 0xc8 is in use    <-- Unknown Device
Notice:  Serial Log:ID 0xa3 is in use    <-- iAqualink 2.0
Notice:  Serial Log:ID 0x68 is in use    <-- LXi / LRZ Heater
Notice:  Serial Log:ID 0x78 is in use    <-- Jandy VSP ePump
```

0x33 is a Aqualink Touch device.
0xa3 is the iAqualink 2 wifi controller. (It also uses the ID above). Webpage is 0x33, Mobile App is 0xa3
0x68 is a Jandy heater
0x78 is a Jandy VSP.
0xc8 I believe this is the UI on the actual panel.  From the images on the Z4 documentation I can tell the panel UI is very similar to a PDA protocol but it is different, so I don't think it's using ID 0x60.

**Reply by zeropoint46** [reply]
Every time I have run the logger 0x60 shows available so I'm not sure what else I can do or change. Of you have suggestions I'm willing to try

**Reply by zeropoint46** [reply]
I mean serial logger has this every time:
```Notice:  Serial Log:ID 0x60 is not used  <-- can use for Aqualinkd (PDA mode only)```
I've restarted aqualinkd multiple times and waited for it to timeout waiting for the 0x60 probe, so not sure what other options or things I can change/adjust.

**Reply by sfeakes** [reply]
Does the UI on the control panel go to sleep / display nothing at any point in time?   If so, maybe it is using the PDA 0x60 ID, and 0xc8 when you run serial_logger it's catching the display in sleep mode.  The PDA does something similar, but will completely disconnect and send a probe message, maybe the Z4 is set to never really disconnect.  I have a new dev release of AqualinkD that *might* help, but if this is the case it may also cause a lot of issues when the panel UI wakes.   I'll let you know once I've posted.

**Reply by sfeakes** [reply]
I just posted 2.4.1 (Dev 0.2). Not a release, just a dev build.  If you try that and when starting AqualinkD add a command line switch `-nc` see if it actually starts and does something.  BUT, I am sure this will interfere with the panel UI, so be careful.  
EDIT
(careful, meaning you might get some strange behavior, it won't damage the panel)

**Reply by zeropoint46** [reply]
Okay, I'll give it a shot either later tonight or tomorrow. From what I can see the display never sleeps, just a static display. If you're in a menu or something like that and leave it, it will go back to the main screen/menu, but never turn off. I don't really care if the NC option makes things go funky, I'll try anything.

**Reply by zeropoint46** [reply]
Okay, I messed around with it a bit with the new build and -nc. Basically it now just sits there without the aqualinkd process dying from timeout of no probe, which I'm guessing is what the -nc flag does. However, nothing works, I can't do anything. I can grab some logs for you later but I'm not sure I saw anything that stuck out. I did see your notes about this new dev version has some prelim support for PDA with iaqualink devid 0x33. I gave that a shot and disconnected my iq20 from the system. Aqualinkd didn't die and I didn't need the -nc flag, however, it couldn't control or read anything.

Maybe I'll just buy some used RS equipment and replace mine? I'm guessing hte JVA's and relays are all compatible? I would just need the powever center metal "box" and a RS cpu board? I'm guessing I could also program it with aqualinkd as well and not even need any touch screens or button panels? Thanks for the input and help.

**Reply by zeropoint46** [reply]
Ohh and if I am able and do upgrade to an RS, I'm more than willing to donate my Z4 board to this project if it helps to get things working for other people. :)

**Reply by sfeakes** [reply]
The -nc = No Check.  Basically tell AqualinkD not to do any startup checks and act on any messages it receives and understands.  So this probably means the Z4 is using the PDA ID 0x60 differently to a PDA panel.  If you put AqualinkD in serial debug mode and use the Panel UI and iq20 interfaces, then post the output, I should be able to decode what’s going on.
PDA with id 0x33 is still under dev, but most things work in the current build, so surprised that you didn’t get anything to work with the iq20 disconnected and current dev build of AqualinkD.  Again, posting serial debug log using those settings will help understand what’s happening.  I’d be surprised if we couldn’t get it working in this mode.

As for upgrading the Z4 to RS, I’m not sure exactly what’s needed and can be reused.  Obviously the RS board is needed (the metal box if you want it to look neat), I would assume the relays are “standard” but not sure. JVA’s will be the same. Power supply is also not cheap for the RS box and not sure if Z4 is same.

EDIT.  Thanks for the offer of the Z4, let’s hope it doesn’t get to that point. But if it does it would be good to use for decoding.

**Reply by sfeakes** [reply]
@zeropoint46   when / if you put AqualinkD in serial debug mode, and cycle through using panel UI etc. can you change the pump RPM from the phone app (not the web based AqualinkTouch), that will force a both protocols of the iq20 to be used.

**Reply by zeropoint46** [reply]
@sfeakes I'll give it a shot, however I can't really directly control my pump rpm from the app. I do have like an aux3 mode and cleaner mode that from what I can tell, adjust the VSP rpm, however I don't actually have direct control of the pump speed/rpm. Hopefully that makes sense. If you want to connect I would be willing to setup some kind of live remote session if you want to get in a debug anything.

**Reply by sfeakes** [reply]
@zeropoint46    Then don't worry about changing RPM.   That's really odd that you can't change the RPM from the mobile App or web app. The iAqualink2/3 connected to an RS panel can. So probably more arbitory limitations they inflicted on the Z4 (like they did the old PDA panels)

**Comment by jswilson64** [comment]
I'm another Z4 owner who's interested in using AqualinkD. Our setup is a pool and spa, heater, a couple of auto-actuated valves to switch pool/spa modes.  It also has an ancient iAqualink IQ900 which I've read only supports 802.11b ?  We've never set it up (owned the house for about 6 months now).

Our Z4 interface is beginning to crap out. We had the main board replaced b/c our pool guy thought that was the problem. Chasing possible power issues (maybe the transformer). 

Looking at AqualinkD as a possible way to get by if we can pull the Z4 interface and the IQ 900.

**Reply by zeropoint46** [reply]
Looks like I better get back to testing and provide more logs to @sfeakes. I'll see if I can find some time this week. Also, not sure if it helps at all, but I have hardwired my iq module to cat5 ethernet so it's not on wifi. Might be an option for you and might make it more stable in the mean time.

**Comment by zeropoint46** [comment]
@sfeakes I never did get this working, however I did end up just dumping my z4 and getting a used RS4 installed so now it's not really an issue for me. That said, I was wondering if you're interested I would donate my z4 to you if you might want to use it to test and develop further. Might be helpful for other people. let me know if you're interested.

**Reply by sfeakes** [reply]
@zeropoint46 Thanks for the offer I will take you up on that, I have a lot of old hardware people have donated over the years that does really help development. Shoot me an email sfeakes@gmail.com
